### ISSUE TITLE
PORT_ERROR After Upgrade to v1.2.2

### ISSUE THREAD

**Issue by bleppard** [zerotier/ZeroTierOne#467]
On a machine that previously ran ZeroTier correctly. I am now receiving a `PORT_ERROR` trying to join a network.

When joining I receive this error:
![image](https://cloud.githubusercontent.com/assets/2224704/24111661/92b667dc-0d6d-11e7-867f-1ccac3ef2622.png)
But the ZeroTier service is running:
![image](https://cloud.githubusercontent.com/assets/2224704/24111708/b9330faa-0d6d-11e7-8bee-28bb3aeecc6b.png)
`listpeers` shows the status as `PORT_ERROR`

I have tried uninstalling / rebooting / installing many times. With no appreciable effect.

Don't know if this is relevant but, I noticed that the zttap300 driver is listed as an unknown device and doesn't have any drivers listed in the driver tab. 
![image](https://cloud.githubusercontent.com/assets/2224704/24111464/d76edf40-0d6c-11e7-8a3d-b7b964a5afab.png)

The following is the output of `info`, `listnetworks` and `listpeers`:
```
C:\Program Files (x86)\ZeroTier\One>zerotier-cli.bat info
200 info c20921896a 1.2.2 ONLINE

C:\Program Files (x86)\ZeroTier\One>zerotier-cli.bat listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 8056c2e21c923a8f Leppard 8e:f8:9b:3d:6b:a8 PORT_ERROR PRIVATE  192.168.140.117/24

C:\Program Files (x86)\ZeroTier\One>zerotier-cli.bat listpeers
200 listpeers <ztaddr> <path> <latency> <version> <role>
200 listpeers 8056c2e21c 37.59.34.61/19993;5534;15302;1.00 218 1.1.19 LEAF
200 listpeers 8841408a2e 45.32.198.130/9993;7044;7044;1.00 274 1.1.5 PLANET
200 listpeers 9d219039f3 159.203.97.171/9993;4856;4808;1.00 332 1.1.5 PLANET
200 listpeers b1d366e81f - 80 1.2.2 LEAF
200 listpeers dcf2d3aed7 - 0 - LEAF
200 listpeers e300a2f735 - 0 - LEAF
```

Attached are the events for the driver, I noticed that even when uninstalled, there were several zttap300 related unknown drivers listed in Device Manager. So I uninstalled all of them and reinstalled ZeroTier, this had no appreciable effect.
[zttap300_events.zip](https://github.com/zerotier/ZeroTierOne/files/855727/zttap300_events.zip)

Running
Windows 10 Pro 10.0.14393
And not running any anti-virus except the built in Windows Defender.

**Comment by bleppard** [comment]
Ok, going back to previous version of ZeroTier didn't work.

But the issue resolved when I manually installed the zttap300.inf by right clicking and choosing Install.

I even tried another uninstall / reboot / reinstall this also worked without a hitch. I have no idea why I had to install the driver manually to get it so successfully install.

On thing that I noticed is that I have trust drivers from Zero Tier checked, so I don't get the pop-up to give authorization to install the driver during the setup process, but I did get that pop-up when I installed manually.

**Comment by adamierymenko** [comment]
Did this happen during or after the auto-update or did it happen after a manual update?

**Comment by bleppard** [comment]
Manual update from v1.1.14.

**Comment by adamierymenko** [comment]
Thanks. We will check it out.

**Comment by gsomoza** [comment]
Bump - same thing but on macOS Sierra 10.12.6

**Comment by alax** [comment]
Yup, issue is still present on the latest Sierra (10.12.6) and ZT (1.2.4).

**Comment by ddhhz** [comment]
@alax rebooting fixed it for me on Sierra (10.12.6)

**Comment by gsomoza** [comment]
Rebooting didn't help me. Also ZT One 1.2.4

**Comment by nindaloo** [comment]
Same issue here.  Sierra (10.12.6) with ZT 1.2.4. -- although it was happening with a prior version...and I updated to see if it would fix but it persists. 

I've tried fresh install, rebooting, etc.  I've confirmed the same issue with others on our network.  PORT_ERROR for all connections.

**Comment by anadin** [comment]
I'm having the issue too, 10.13.0 with ZT 1.2.4, not having it on another 10.13 iMac.

**Comment by anadin** [comment]
More info...

running this script 

https://github.com/zerotier/ZeroTierOne/blob/0daff26fba640f2ef167ad66974dc1c4a6c79c01/ext/installfiles/mac/postinst.sh

 brings up a dialogue saying unable to connect to network.

The tap kext seems to not be loadable
`sudo kextutil tap.kext
Untrusted kexts are not allowed
Kext rejected due to system policy: <OSKext 0x7fb01e92eed0 [0x7fffa6e0f5b0]> { URL = "file:///Library/StagedExtensions/Library/Application%20Support/ZeroTier/One/tap.kext/", ID = "com.zerotier.tap" }
Diagnostics for tap.kext:`

**Comment by anadin** [comment]
OK so if anyone ever finds this with 10.13 (High Sierra) you have 30mins after install to approve the KEXT.

http://blog.eriknicolasgomez.com/2017/07/25/Kextpocalypse-High-Sierra-and-kexts-in-the-Enterprise/

**Comment by cormacrelf** [comment]
Same thing on Windows 10 1703, on a clean install, manually installing `C:\Program Files\ZeroTier\One\zttap300.inf` worked.

I did run the installer silently the first time, and don't remember approving a driver. Subsequent remove/reinstalls did not show the driver approval either.

**Comment by gsomoza** [comment]
I finally figured it out, after several months. For those running into a similar issue in Mac OSX: you may already be running a different `tap` kext that's preventing `com.zerotier.tap` from starting.

Here's how you can diagnose and fix (assuming you have the same problem as me): 

1) Open a terminal and try to load the ZeroTier Tap kext by running `sudo kextutil -v 6 /Library/Application\ Support/ZeroTier/One/tap.kext`. In your output you should see something like this:

```
...
Loading /Library/StagedExtensions/Library/Application Support/ZeroTier/One/tap.kext.
...
(kernel) Loading kext com.zerotier.tap.
...
(kernel) Kext com.zerotier.tap executable loaded; 9 pages at 0xffffff7f84017000 (load tag 198).
(kernel) Kext com.zerotier.tap calling module start function.
(kernel) Kext com.zerotier.tap did not start (return code 0x5).
(kernel) Kext com.zerotier.tap start failed (result 0x5).
(kernel) Kext com.zerotier.tap failed to load (0xdc008017).
(kernel) Failed to load kext com.zerotier.tap (error 0xdc008017).
(kernel) Kext com.zerotier.tap unloading.
...
```

In other words: the text executable was actually loaded but it failed to start after calling the module start function, so it was then unloaded.

2) To see which other tap kexts you may be running you can run `kextstat | grep tun`, which in my case showed that kext `net.sf.tuntaposx.tun` was running.

3) If you feel confident about what you're doing (you may break some running application?!) then you can unload the kext that's giving you a conflict. In my case: `sudo kextunload -b  net.sf.tuntaposx.tun`

4) Now proceed to issue the command to load the ZeroTier kext again (step 1) and this time it should work.

I'm not sure what happens after rebooting. What I'll do next is see to which application/service `net.sf.tuntaposx.tun` belongs to so I can hopefully uninstall it completely.

**Comment by faddat** [comment]
```
admins-MacBook-Pro:blockchain admin$ sudo kextutil -v 6 /Library/Application\ Support/ZeroTier/One/tap.kext
Password:
Kext user-space log filter changed from 0xff2 to 0xfff.
Kext kernel-space log filter changed from 0xff2 to 0xfff.
Kext library architecture set to x86_64.
Can't stat /AppleInternal - No such file or directory.
Defaulting to kernel file '/System/Library/Kernels/kernel'
Kext library recording diagnostics for: validation authentication dependencies warnings.
Reading loaded kext info from kernel.
(kernel) User-space log flags changed from 0x0 to 0xfff.
(kernel) Received 'Get Loaded Kext Info' request from user space.
(kernel) Returning loaded kext info.
Loaded kext info:
{
}

Kext rejected due to system policy: <OSKext 0x7fda5cc0d430 [0x7fff916bd980]> { URL = "file:///Library/StagedExtensions/Library/Application%20Support/ZeroTier/One/tap.kext/", ID = "com.zerotier.tap" }
Diagnostics for /Library/Application Support/ZeroTier/One/tap.kext:
```

**Comment by victorhooi** [comment]
I have machine that seems to be in the same boat (user probably didn't accept the Kext within 30 minutes).

```
bash-3.2$ sudo kextutil -v 6 /Library/Application\ Support/ZeroTier/One/tap.kext 
Password:
Sorry, try again.
Password:
Kext user-space log filter changed from 0xff2 to 0xfff.
Kext kernel-space log filter changed from 0xff2 to 0xfff.
Kext library architecture set to x86_64.
Can't stat /AppleInternal - No such file or directory.
Defaulting to kernel file '/System/Library/Kernels/kernel'
Kext library recording diagnostics for: validation authentication dependencies warnings.
Reading loaded kext info from kernel.
(kernel) User-space log flags changed from 0x0 to 0xfff.
(kernel) Received 'Get Loaded Kext Info' request from user space.
(kernel) Returning loaded kext info.
Loaded kext info:
{
}

Kext rejected due to system policy: <OSKext 0x7feb9ad21690 [0x7fff85be8af0]> { URL = "file:///Library/StagedExtensions/Library/Application%20Support/ZeroTier/One/tap.kext/", ID = "com.zerotier.tap" }
Diagnostics for /Library/Application Support/ZeroTier/One/tap.kext:
bash-3.2$ 
```
Any ideas on how to resolve this?

**Comment by anadin** [comment]
On macos if you reboot a button should appear in security preferences for a while allowing you to approve it

**Comment by ydakuka** [comment]
@faddat @victorhooi https://github.com/tekezo/Karabiner-Elements/issues/777#issuecomment-312400158

**Comment by osdiab** [comment]
I am getting this error as well, I have already allowed the kext install but get the same error from @gsomoza 's message:

```
→ sudo kextutil -v 6 /Library/Application\ Support/ZeroTier/One/tap.kext
Kext user-space log filter changed from 0xff2 to 0xfff.
Kext kernel-space log filter changed from 0xff2 to 0xfff.
Kext library architecture set to x86_64.
Can't stat /AppleInternal - No such file or directory.
Defaulting to kernel file '/System/Library/Kernels/kernel'
Kext library recording diagnostics for: validation authentication dependencies warnings.
Reading loaded kext info from kernel.
(kernel) User-space log flags changed from 0x0 to 0xfff.
(kernel) Received 'Get Loaded Kext Info' request from user space.
(kernel) Returning loaded kext info.
Loaded kext info:
{
}

/Library/Application Support/ZeroTier/One/tap.kext appears to be loadable (not including linkage for on-disk libraries).
Reading loaded kext info from kernel.
(kernel) User-space log flags changed from 0x0 to 0xfff.
(kernel) Received 'Get Loaded Kext Info' request from user space.
(kernel) Returning loaded kext info.
Loaded kext info:
{
}

Loading /Library/Application Support/ZeroTier/One/tap.kext.
Reading load info for 5 kexts.
Reading loaded kext info from kernel.
(kernel) User-space log flags changed from 0x0 to 0xfff.
(kernel) Received 'Get Loaded Kext Info' request from user space.
(kernel) Returning loaded kext info.
Loaded kext info:
{
"com.apple.kpi.libkern" = {
    "OSBundleWiredSize" = 54320
    "CFBundleIdentifier" = "com.apple.kpi.libkern"
    "OSBundleUUID" = <376e1a00 66974ee5 ade593b3 87fce2ad>
    "OSBundleLoadAddress" = -551892770816
    "OSBundleLoadTag" = 4
    "OSBundleStarted" = true
    "OSBundleCompatibleVersion" = "8.0.0d0"
    "OSKernelResource" = true
    "CFBundleVersion" = "17.6.0"
    "OSBundlePath" =
        "/System/Library/Extensions/System.kext/PlugIns/Libkern.kext"
    "OSBundleRetainCount" = 174
    "OSBundlePrelinked" = true
    "OSBundleIsInterface" = true
    "OSBundleLoadSize" = 54320
}
"com.apple.kpi.unsupported" = {
    "OSBundleWiredSize" = 24224
    "CFBundleIdentifier" = "com.apple.kpi.unsupported"
    "OSBundleUUID" = <786b4094 bf6941fa a964ad9b ff9c2d3b>
    "OSBundleLoadAddress" = -551892230144
    "OSBundleLoadTag" = 7
    "OSBundleStarted" = true
    "OSBundleCompatibleVersion" = "8.0.0b1"
    "OSKernelResource" = true
    "CFBundleVersion" = "17.6.0"
    "OSBundlePath" =
        "/System/Library/Extensions/System.kext/PlugIns/Unsupported.kext"
    "OSBundleRetainCount" = 95
    "OSBundlePrelinked" = true
    "OSBundleIsInterface" = true
    "OSBundleLoadSize" = 24224
}
"com.apple.kpi.mach" = {
    "OSBundleWiredSize" = 16224
    "CFBundleIdentifier" = "com.apple.kpi.mach"
    "OSBundleUUID" = <04ef86f1 2a40471c 864ba52e 51d09924>
    "OSBundleLoadAddress" = -551892713472
    "OSBundleLoadTag" = 5
    "OSBundleStarted" = true
    "OSBundleCompatibleVersion" = "8.0.0d0"
    "OSKernelResource" = true
    "CFBundleVersion" = "17.6.0"
    "OSBundlePath" = "/System/Library/Extensions/System.kext/PlugIns/Mach.kext"
    "OSBundleRetainCount" = 153
    "OSBundlePrelinked" = true
    "OSBundleIsInterface" = true
    "OSBundleLoadSize" = 16224
}
"com.apple.kpi.bsd" = {
    "OSBundleWiredSize" = 40496
    "CFBundleIdentifier" = "com.apple.kpi.bsd"
    "OSBundleUUID" = <6612a6c8 b11c4736 9c1012c7 73b9340f>
    "OSBundleLoadAddress" = -551892811776
    "OSBundleLoadTag" = 1
    "OSBundleStarted" = true
    "OSBundleCompatibleVersion" = "8.0.0b1"
    "OSKernelResource" = true
    "CFBundleVersion" = "17.6.0"
    "OSBundlePath" =
        "/System/Library/Extensions/System.kext/PlugIns/BSDKernel.kext"
    "OSBundleRetainCount" = 133
    "OSBundlePrelinked" = true
    "OSBundleIsInterface" = true
    "OSBundleLoadSize" = 40496
}
}

/System/Library/Extensions/System.kext/PlugIns/Libkern.kext (version 17.6, UUID 376E1A00-6697-4EE5-ADE5-93B387FCE2AD) is loaded.
/System/Library/Extensions/System.kext/PlugIns/Libkern.kext (version 17.6): is started.
setting load address of /System/Library/Extensions/System.kext/PlugIns/Libkern.kext to 0xffffff7f80a0a000
/System/Library/Extensions/System.kext/PlugIns/Unsupported.kext (version 17.6, UUID 786B4094-BF69-41FA-A964-AD9BFF9C2D3B) is loaded.
/System/Library/Extensions/System.kext/PlugIns/Unsupported.kext (version 17.6): is started.
setting load address of /System/Library/Extensions/System.kext/PlugIns/Unsupported.kext to 0xffffff7f80a8e000
/System/Library/Extensions/System.kext/PlugIns/Mach.kext (version 17.6, UUID 04EF86F1-2A40-471C-864B-A52E51D09924) is loaded.
/System/Library/Extensions/System.kext/PlugIns/Mach.kext (version 17.6): is started.
setting load address of /System/Library/Extensions/System.kext/PlugIns/Mach.kext to 0xffffff7f80a18000
/System/Library/Extensions/System.kext/PlugIns/BSDKernel.kext (version 17.6, UUID 6612A6C8-B11C-4736-9C10-12C773B9340F) is loaded.
/System/Library/Extensions/System.kext/PlugIns/BSDKernel.kext (version 17.6): is started.
setting load address of /System/Library/Extensions/System.kext/PlugIns/BSDKernel.kext to 0xffffff7f80a00000
Omitting loaded kext /System/Library/Extensions/System.kext/PlugIns/Libkern.kext from mkext for kernel load.
Omitting loaded kext /System/Library/Extensions/System.kext/PlugIns/Unsupported.kext from mkext for kernel load.
Omitting loaded kext /System/Library/Extensions/System.kext/PlugIns/Mach.kext from mkext for kernel load.
Omitting loaded kext /System/Library/Extensions/System.kext/PlugIns/BSDKernel.kext from mkext for kernel load.
Adding /Library/StagedExtensions/Library/Application Support/ZeroTier/One/tap.kext to mkext.
/Library/StagedExtensions/Library/Application Support/ZeroTier/One/tap.kext added 50496-byte noncompressed executable to mkext.
Created mkext for architecture x86_64 containing 1 kexts.
Loading /Library/StagedExtensions/Library/Application Support/ZeroTier/One/tap.kext.
(kernel) User-space log flags changed from 0x0 to 0xfff.
(kernel) Received kext load request from user space.
(kernel) Received request from user space to load kext com.zerotier.tap.
(kernel) Loading kext com.zerotier.tap.
(kernel) Kext com.apple.kpi.libkern is already loaded.
(kernel) Kext com.apple.kpi.unsupported is already loaded.
(kernel) Kext com.apple.kpi.mach is already loaded.
(kernel) Kext com.apple.kpi.bsd is already loaded.
(kernel) Allocated link buffer for kext com.zerotier.tap at 0xffffff7fa200a000 (36864 bytes).
(kernel) Kext com.zerotier.tap executable loaded; 9 pages at 0xffffff7f83a0a000 (load tag 197).
(kernel) Kext com.zerotier.tap calling module start function.
(kernel) Kext com.zerotier.tap did not start (return code 0x5).
(kernel) Kext com.zerotier.tap start failed (result 0x5).
(kernel) Kext com.zerotier.tap failed to load (0xdc008017).
(kernel) Failed to load kext com.zerotier.tap (error 0xdc008017).
(kernel) Kext com.zerotier.tap unloading.
(kernel) Kext com.zerotier.tap unwiring and unmapping linked executable.
(kernel) Kext com.zerotier.tap unloaded.
(kernel) Kext com.zerotier.tap removing all personalities naming it from the IOCatalogue.
Kernel error handling kext request - (libkern/kext) kext (kmod) start/stop routine failed.
Failed to load /Library/StagedExtensions/Library/Application Support/ZeroTier/One/tap.kext - (libkern/kext) kext (kmod) start/stop routine failed.
Failed to load /Library/Application Support/ZeroTier/One/tap.kext - (libkern/kext) kext (kmod) start/stop routine failed.
```

I tried reinstalling it completely and nuking the `/Library/Application\ Support/ZeroTier/One` directory but that didn't help. Stuck :(

**Comment by laduke** [comment]
Some people have had luck with deleting `/Library/Application\ Support/ZeroTier/One/tap.kext`

and then rebooting, re-installing, etc

**Comment by osdiab** [comment]
Did that, no change :P

**Comment by wanjochan** [comment]
seems problem remain unsolved on OSX....

**Comment by wanjochan** [comment]
Here good new after google a lots:
```
# Open "Security and Privacy" at the OSX/Settings:
open -b com.apple.systempreferences /System/Library/PreferencePanes/Security.prefPane
```
then click ALLOW the zerotier....

and more, maybe later some one will get the command by sudo directly...
sudo /usr/libexec/PlistBuddy ~/Library/Preferences//com.zerotier.ZeroTier-One.plis -c "Set: ???"

**Comment by wanjochan** [comment]
Here my steps for your fellows reference:

* sudo launchctl unload /Library/LaunchDaemons/com.zerotier.one.plist
* sudo launchctl load /Library/LaunchDaemons/com.zerotier.one.plist
* open "Security and Privacy" and try click Allow ZeroTier /* maybe no effects....nevermind, just continue */
* REBOOT the OSX  /* VERY IMPORTANT */
* open "Security and Privacy" and try click Allow ZeroTier AGAIN /* if it still there */
* sudo launchctl unload /Library/LaunchDaemons/com.zerotier.one.plist
* sudo launchctl load /Library/LaunchDaemons/com.zerotier.one.plist
* zerotier-cli listnetworks /* now the PORT_ERROR should be gone */
* ping your nodes on the network should now being OK

**Comment by SwiTool** [comment]
That worked @wanjochan thanks

**Comment by lijiansong** [comment]
> Bump - same thing but on macOS Sierra 10.12.6

I have the same thing but on macOS Sierra 10.13.1

**Comment by simpleapples** [comment]
@wanjochan thx, it works!

**Comment by wanjochan** [comment]
P.S.  command line for opening "Security and Privacy"

```
open "x-apple.systempreferences:com.apple.preference.security?General"
```

**Comment by dalegaspi** [comment]
i'm on Mojave and running into this problem as of this writing.  i tried the [tip from @wanjochan|https://github.com/zerotier/ZeroTierOne/issues/467#issuecomment-426852518] to no avail.  the prompt to Allow Zero Tier is not even appearing in the Security and Privacy Pref Pane.

**Comment by HunterDG** [comment]
On Mojave, I noticed that I still had "Port Error" after authorizing the kext in System Prefs.  

If using ZeroTier One (GUI) Make sure to **select the desired network** (so it has a checkmark by it).  Mine was not selected.  Not sure if I did that accidentally or if it deselects itself after connection failure due to unauthorized kext.

**Comment by dalegaspi** [comment]
@HunterDG that didn't work for me, either.

**Comment by witer55** [comment]
not work on the MacBook Air (M1, 2020) macOS 12.1 Monterey, it still PORT_ERROR on Mac

@jade-42

**Comment by nberanger** [comment]
These instructions are out of date. I have everything running, but there is no kernel extension or system extension installed.
